src/BenchmarksApps/TodosApi/DataExtensions.cs

```diff
@@ -0,0 +1,298 @@
+﻿using System.Data;
+
+namespace Npgsql;
+
+internal static class DataExtensions
+{
+    public static async ValueTask OpenIfClosedAsync(this NpgsqlConnection connection)
+    {
+        if (connection.State == ConnectionState.Closed)
+        {
+            await connection.OpenAsync();
+        }
+    }
+
+    public static async Task<int> ExecuteAsync(this NpgsqlConnection connection, string commandText)
+    {
+        await using var cmd = connection.CreateCommand(commandText);
+
+        await connection.OpenIfClosedAsync();
+        return await cmd.ExecuteNonQueryAsync();
+    }
+
+    public static async Task<int> ExecuteAsync(this NpgsqlDataSource dataSource, string commandText)
+    {
+        await using var cmd = dataSource.CreateCommand(commandText);
+
+        return await cmd.ExecuteNonQueryAsync();
+    }
+
+    public static async Task<object?> ExecuteScalarAsync(this NpgsqlDataSource dataSource, string commandText)
+    {
+        await using var cmd = dataSource.CreateCommand(commandText);
+
+        return await cmd.ExecuteScalarAsync(CancellationToken.None);
+    }
+
+    public static async Task<int> ExecuteAsync(this NpgsqlConnection connection, string commandText, params NpgsqlParameter[] parameters)
+    {
+        await using var cmd = connection.CreateCommand(commandText, parameters);
+
+        await connection.OpenIfClosedAsync();
+        return await cmd.ExecuteNonQueryAsync();
+    }
+
+    public static async Task<int> ExecuteAsync(this NpgsqlDataSource dataSource, string commandText, params NpgsqlParameter[] parameters)
+    {
+        await using var cmd = dataSource.CreateCommand(commandText, parameters);
+
+        return await cmd.ExecuteNonQueryAsync();
+    }
+
+    public static async Task<int> ExecuteAsync(this NpgsqlConnection connection, string commandText, Action<NpgsqlParameterCollection> configureParameters)
```

Comment:
I didn't think we were using `NpgsqlConnection` anymore. It doesn't look like a bunch of these overloads are used. Can we delete any methods that aren't being used?

At least all the `NpgsqlConnection` methods can be removed since we use DataSource API now.